src/cgutils.cpp

```diff
@@ -3547,8 +3595,17 @@ static jl_cgval_t emit_new_struct(jl_codectx_t &ctx, jl_value_t *ty, size_t narg
             for (unsigned i = 0; i < na; i++) {
                 jl_value_t *jtype = jl_svecref(sty->types, i); // n.b. ty argument must be concrete
                 jl_cgval_t fval_info = argv[i];
+
+                IRBuilderBase::InsertPoint savedIP;
                 emit_typecheck(ctx, fval_info, jtype, "new");
                 fval_info = update_julia_type(ctx, fval_info, jtype);
+                // TODO: Use (post-)domination instead.
+                bool field_promotable = !init_as_value && fval_info.promotion_ssa != -1 &&
+                    fval_info.promotion_point && fval_info.promotion_point->getParent() == ctx.builder.GetInsertBlock();
```

Comment:
There may also need to be a check here that the type being stored here is compatible, since we might not post-dominate in execution (via a typeassert call), even though we observe post-domination in IR order (yes, this is also a problem with our current errors violating dominance)? The `update_julia_type` will partly help with that already (by checking for trivial cases where the types are clearly incompatible, but might miss a case where we store a concrete struct into a field–e.g. `Int`–that is a union of multiple other types which excludes our type–e.g. `Union{Nothing,Int8}`).

Reply:
Yeah, I'm struggling with this one too :/ so I defer to @Keno. But what I got is that a typeassert might make us do an incompatible store but we don't see that because it doesn't dominate the IR we are locally looking at?

Reply:
Hypothetically, the store might be post-dominated in the normal flow of the IR, but be removed by an exception in the runtime flow